### PR TITLE
Standalone ZMQ-Encoder

### PR DESCRIPTION
**Pull request by joelosw**
Ticket: [AK-397](https://psiori.atlassian.net/browse/AK-397)

Trying to implement a solution that receives zmq messages from code and uses AVTransmitter to encode those and send them out via rtp.

Currently prototype is working, but still having troubles with CMake.
Also, hardcoded values (hosts, ports, ...) need to be changed to cmd-line arguments

### CODE REVIEW

**Comment by joelosw** [comment]
Can confirm, that is working locally.
Unfortunately not yet able to test on working system because minicrane down and palatka network restrictions. 
However, finding tclap is currently ownly a workaround, I have not figured out how to get CMake to find it properly, even though it does work in autocrane-core 🤔

File: CMakeLists.txt [hunk 1]
```diff
@@ -2,60 +2,89 @@ project(zmq-streaming C CXX)
...
 add_compile_options(-fno-lto)
-
 option(WITH_SPINNAKER "Use spinnaker?" OFF)
 
-find_package(Boost COMPONENTS thread REQUIRED)
-message(STATUS "Boost_LIBRARIES ${Boost_LIBRARIES}")
+# To fix OpenMP not found -> use brews llvm clang, not Apple Clang.
+# Taken from https://stackoverflow.com/a/53889826
+cmake_policy(GET CMP0025 InfoCompile)
+message(STATUS "CMP0025 ${InfoCompile}")
```
Comment (themightyoarfish) [inline CMakeLists.txt:14]: should be obsolete with new opencv install, right?

File: CMakeLists.txt [hunk 2]
```diff
@@ -2,60 +2,89 @@ project(zmq-streaming C CXX)
...
 
+list(INSERT CMAKE_PREFIX_PATH 0 ${AUTOCRANE_DEPS})
+set(Boost_ROOT ${AUTOCRANE_DEPS})
+find_package(Boost COMPONENTS thread REQUIRED PATHS)
+message(STATUS " Boost_INCLUDE_DIRS: ${Boost_INCLUDE_DIRS} ")
+message(STATUS " Boost_LIBRARIES ${Boost_LIBRARIES} ")
+set(OpenCV_ROOT ${AUTOCRANE_DEPS})
 find_package(OpenCV COMPONENTS core highgui imgcodecs imgproc REQUIRED)
```
Comment (themightyoarfish) [inline CMakeLists.txt:21]: you should be able to just pass -DCMAKE_PREFIX_PATH=<deps> if you use the recent upstream master

File: encode_from_zmq.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+#include "avtransmitter.hpp"
+#include "time_functions.hpp"
+
+using std::string;
+using namespace std::chrono;
+using namespace TCLAP;
+
+class VideoStreamMonitor {
+private:
+  AVTransmitter* transmitter;
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:23]: no need for pointer

File: encode_from_zmq.cpp [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+   * resized. For value: 0, crop is returned as it is. For val: 1, crop is
+   * resized to original frame size.
+   */
+  double output_scale_factor;
+
+  /**
+   * @brief Enable or disable image zooming functionality. If true, images can
+   * be zoomed in or out via TUI.
+   */
+  bool do_zoom;
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:40]: can go away, as there is no tui

File: encode_from_zmq.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  socket.set(zmq::sockopt::plain_password, pw);
+  socket.set(zmq::sockopt::rcvhwm, 2);
+  socket.set(zmq::sockopt::subscribe, topic);
+  const auto connect_str = string("tcp://") + host + ":" + std::to_string(port);
+  socket.connect(connect_str);
+  std::cout << "Connected zmq socket to " << connect_str
+            << ", on topic: " << topic << std::endl;
+  std::cout << "Starting main" << std::endl;
+
+  auto streamer = new VideoStreamMonitor(
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:188]: no need for pointer

File: encode_from_zmq.cpp [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  while (true) {
+    std::cout << "Waiting for zmq message" << std::endl;
+    result = socket.recv(recv_topic, zmq::recv_flags::none);
+    std::cout << "Got topic, waiting for zmq result" << std::endl;
+    result                = socket.recv(request, zmq::recv_flags::none);
+    const string topicStr = recv_topic.to_string();
+    std::cout << "Received zmq message on topic " << topicStr << std::endl;
+    string requestStr = request.to_string();
+    if (topicStr.find("camera|") != string::npos) {
+      int64_t more = socket.get(zmq::sockopt::rcvmore);
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:205]: unused?

File: encode_from_zmq.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+      uchar* ptr   = reinterpret_cast<uchar*>(request.data());
+      std::vector<uchar> data(ptr, ptr + request.size());
+      cv::Mat image = cv::imdecode(data, -1);
+      streamer->process(image);
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:209]: for this application we probably want a raw image option for the zmq monitor, otherwise the image will just get encoded twice and add to the delay. if that exists already, i would use it, or use ppm

File: encode_from_zmq.cpp (hunk 4 above)
Comment (themightyoarfish) [reply encode_from_zmq.cpp:40]: seems like youre supporting this as cmd line args. i guess its fine, although this functionality was never used as it was made for wolffkran

File: encode_from_zmq.cpp [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    do_zoom(do_zoom),
+    has_printed_sdp(false) {
+  stop.store(false);
+  av_log_set_level(AV_LOG_QUIET);
+  std::cout << "Streaming to " << host << ":" << port << std::endl;
+  encoder = std::thread([&]() {
+    while (!stop.load()) {
+      // const auto queue_status = queue.try_pull_front(image);
+      //  keep the zoom_factor value in range [0.2, 1.0]
+      cv::Mat image = queue.pull_front();
```
Comment (themightyoarfish) [inline encode_from_zmq.cpp:74]: does this block forever? can it happen that you cant quit the program properly when core doesnt send anything?

Comment (joelosw) [reply encode_from_zmq.cpp:74]: Tried it out. It does block the loop until new image arrives. But you can still quit with ^C

File: encode_from_zmq.cpp (hunk 7 above)
Comment (joelosw) [reply encode_from_zmq.cpp:209]: Seems like we don't support raw yet:
[ZMQMonitor.cpp Line 79](https://github.com/psiori/autocrane-core/blob/ff7e0a6dbf214da6a037a20f8c55ec064da030ff/src/Sensors/Monitors/ZMQMonitor.cpp#L79)
I'll go with ppm then

File: encode_from_zmq.cpp (hunk 6 above)
Comment (joelosw) [reply encode_from_zmq.cpp:205]: Yep, more was unused, I removed it. 
I kept the camera filter because sometimes I saw messages of type crane-metric coming in, even though it did not make much sense portwise.... But I guess maybe in local network someone else was running a core with different port-mapping.

File: encode_from_zmq.cpp (hunk 7 above)
Comment (joelosw) [reply encode_from_zmq.cpp:208]: Mhhhm, going from jpg->ppm massively slows down the decoding of the image. 
Without measurements I would guess by a factor of 10

File: CMakeLists.txt (hunk 1 above)
Comment (joelosw) [reply CMakeLists.txt]: yep, right

File: encode_from_zmq.cpp (hunk 3 above)
Comment (joelosw) [reply encode_from_zmq.cpp]: fixed

File: encode_from_zmq.cpp (hunk 5 above)
Comment (joelosw) [reply encode_from_zmq.cpp]: fixed

File: encode_from_zmq.cpp (hunk 4 above)
Comment (joelosw) [reply encode_from_zmq.cpp]: removed it, I was not aware that it was a deprecated feature for wolffkran

File: CMakeLists.txt (hunk 2 above)
Comment (joelosw) [reply CMakeLists.txt]: indeed I can. Thanks!

File: encode_from_zmq.cpp (hunk 6 above)
Comment (themightyoarfish) [reply encode_from_zmq.cpp:171]: ok this seems like a config problem that should be fixed, but i guess it doesnt hurt to check validity of the input.

File: encode_from_zmq.cpp (hunk 7 above)
Comment (themightyoarfish) [reply encode_from_zmq.cpp]: wtf, for real? that seems very strange. sure its more memory, but at least its not decompression

File: avtransmitter.cpp [hunk 9]
```diff
@@ -31,7 +33,7 @@ AVTransmitter::AVTransmitter(const std::string &host, const unsigned int port,
                              avutils::av_strerror2(success));
   }
 
-  this->out_codec = avcodec_find_encoder(AV_CODEC_ID_VP9);
+  this->out_codec = (AVCodec*)avcodec_find_encoder(AV_CODEC_ID_VP9);
```
Comment (themightyoarfish) [inline avtransmitter.cpp:36]: does this build with -DSTRICT=ON? afaik old-style casts should error there.

Comment (joelosw) [reply avtransmitter.cpp]: Mhhm, indeed it does not. I kind of have a sense that I added the casts for a reason (it gave me an error I think), but apparently it also works without them now.... 🤔

File: encode_from_zmq.cpp (hunk 7 above)
Comment (joelosw) [reply encode_from_zmq.cpp]: I am not exactly sure if it is the decoding, but I checked that core is still sending images each cycle at 20Hz, and stream-encoding-log messages suggest that the time is not slower either. So only part left would be the decoding of the incoming zmq-images I guess 🤔

Comment (joelosw) [reply encode_from_zmq.cpp]: But I'll do some tic-tocing to find out

Comment (joelosw) [reply encode_from_zmq.cpp]: Mhhm, oky it is not the cv::imdecode, that is indeed even faster with ppm (~4ms -> ~1ms).
But still, the cycle-frequency of the loop drops from ~20 Hz to ~3Hz with ppm. 
I can't explain why yet

File: CMakeLists.txt [hunk 10]
```diff
@@ -110,63 +110,96 @@ target_sources(decode_rtp PRIVATE
...
-#   # During install, all headers are moved to a include/psiact subdirectory,
+# Install headers into include/${libavstream_LIBRARY}/*, keeping the src directory structure.
+# This matches what other libraries (opencv, tensorflow..) do, although they
+# already maintain the explicit include/LIBRARY_NAME structure in their source
+# tree. We prefer to have the headers next to the sources, so we need some
+# mockery to still allow namespaced includes.
+set(CMAKE_INSTALL_INCLUDEDIR ${CMAKE_INSTALL_PREFIX}/include)
+set(CMAKE_BUILD_INCLUDEDIR ${CMAKE_BINARY_DIR}/include)
+foreach(file ${STREAM_HEADERS})
+  set_target_properties(cppzmq PROPERTIES PUBLIC_HEADER ${file})
```
Comment (ahoereth) [inline CMakeLists.txt:177]: this also needs your library instead of cppzmq, cppzmq was a typo by me

File: CMakeLists.txt [hunk 11]
```diff
@@ -110,63 +110,96 @@ target_sources(decode_rtp PRIVATE
...
+#   # During build time we create the include/${libavstream_LIBRARY}/**/*.h structure within
 #   # the ./build directory, such that it resembles what we have after an install
 #   # without the actual need to move the headers out of the ./src directory.
-#   # message(STATUS "Copy ${file} to ${CMAKE_BUILD_INCLUDEDIR}/psiact/${dir}")
-#   file(MAKE_DIRECTORY ${CMAKE_BUILD_INCLUDEDIR}/psiact/${dir})
-#   configure_file(${file} ${CMAKE_BUILD_INCLUDEDIR}/psiact/${dir} COPYONLY)
-# endforeach()
+#   # message(STATUS "Copy ${file} to ${CMAKE_BUILD_INCLUDEDIR}/${libavstream_LIBRARY}/${dir}")
+  file(MAKE_DIRECTORY ${CMAKE_BUILD_INCLUDEDIR}/${libavstream_LIBRARY}/${dir})
+  configure_file(${file} ${CMAKE_BUILD_INCLUDEDIR}/${libavstream_LIBRARY}/${dir} COPYONLY)
```
Comment (ahoereth) [inline CMakeLists.txt:191]: if we do that you can include the hpp files using the namespace instead of plain #include "header.hpp": `#include "namespace/header.hpp"`

Comment (joelosw) [reply CMakeLists.txt]: 👍 
did that with `"av-streaming/avtransmitter.hpp"` in `VideoStreamMonitor.cpp`

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,9 +1,11 @@
 project(zmq-streaming C CXX)
 cmake_minimum_required(VERSION 3.15)
 set(CMAKE_CXX_STANDARD 14)
-set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+list(APPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR})
 # Use linked libraries in rpath, otherwise binaries in /bin try to call wrong libs in /usr/local etc.
-set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+list(APPEND CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
```
Comment (ahoereth) [inline CMakeLists.txt:8]: CMAKE_INSTALL_RPATH_USE_LINK_PATH is a bool, not a list, no?

Comment (joelosw) [reply CMakeLists.txt:8]: True, got mixed up during Merge Resolve :o
Will fix after the weekend

**Comment by joelosw** [comment]
Any objections to merge?

File: CMakeLists.txt [hunk 13]
```diff
@@ -1,61 +1,207 @@
...
 cmake_minimum_required(VERSION 3.15)
 set(CMAKE_CXX_STANDARD 14)
-set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
-LIST(APPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR})
-add_compile_options(-fno-lto)
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR})
+list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)
+list(APPEND CMAKE_PREFIX_PATH ${CMAKE_CURRENT_LIST_DIR})
+# Use linked libraries in rpath, otherwise binaries in /bin try to call wrong libs in /usr/local etc.
+set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
```
Comment (themightyoarfish) [inline CMakeLists.txt:8]: what does this do? we are not installing any binaries, are we?

File: CMakeLists.txt [hunk 14]
```diff
@@ -1,61 +1,207 @@
...
+# ##############################################################################
+# DETECT SYSTEM  #
+# ##############################################################################
+if(${CMAKE_SYSTEM_NAME} MATCHES Darwin)
+  set(OSX TRUE)
+elseif(${CMAKE_SYSTEM_NAME} MATCHES Linux)
+  set(LINUX TRUE)
+elseif(${WIN32})
+  set(WINDOWS TRUE)
+  include(${CMAKE_CURRENT_LIST_DIR}/tools/master_control/windows/CMakeLists.txt)
```
Comment (themightyoarfish) [inline CMakeLists.txt:18]: why is this project including an unrelated cmake file?

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,61 +1,207 @@
...
+endif()
 
+message(STATUS "stream CMAKE_PREFIX_PATH  ${CMAKE_PREFIX_PATH}")
+add_compile_options(-fno-lto)
```
Comment (themightyoarfish) [inline CMakeLists.txt:22]: Reason for this?

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,61 +1,207 @@
...
 option(WITH_SPINNAKER "Use spinnaker?" OFF)
 
-find_package(Boost COMPONENTS thread REQUIRED)
-message(STATUS "Boost_LIBRARIES ${Boost_LIBRARIES}")
-
+set(Boost_ROOT ${AUTOCRANE_DEPS})
+find_package(Boost COMPONENTS thread REQUIRED PATHS)
```
Comment (themightyoarfish) [inline CMakeLists.txt:26]:
```suggestion
find_package(Boost COMPONENTS thread REQUIRED)
```

File: CMakeLists.txt [hunk 17]
```diff
@@ -1,61 +1,207 @@
...
+endif()
 
+message(STATUS "stream CMAKE_PREFIX_PATH  ${CMAKE_PREFIX_PATH}")
+add_compile_options(-fno-lto)
 option(WITH_SPINNAKER "Use spinnaker?" OFF)
 
-find_package(Boost COMPONENTS thread REQUIRED)
-message(STATUS "Boost_LIBRARIES ${Boost_LIBRARIES}")
-
+set(Boost_ROOT ${AUTOCRANE_DEPS})
```
Comment (themightyoarfish) [inline CMakeLists.txt:25]: will be empty unless defined outside.

Comment (themightyoarfish) [reply CMakeLists.txt:25]: also shouldnt be needed, if you set prefix path appropriately

File: CMakeLists.txt [hunk 18]
```diff
@@ -1,61 +1,207 @@
...
+find_package(Boost COMPONENTS thread REQUIRED PATHS)
+message(STATUS " Boost_INCLUDE_DIRS: ${Boost_INCLUDE_DIRS} ")
+message(STATUS " Boost_LIBRARIES ${Boost_LIBRARIES} ")
```
Comment (themightyoarfish) [inline CMakeLists.txt:28]:
```suggestion
message(STATUS "Boost_INCLUDE_DIRS: ${Boost_INCLUDE_DIRS} ")
message(STATUS "Boost_LIBRARIES ${Boost_LIBRARIES} ")
```

File: CMakeLists.txt [hunk 19]
```diff
@@ -1,61 +1,207 @@
...
 # *must* come after lib A. it does not help if lib B has already been passed previously
-find_package(FFMPEG REQUIRED COMPONENTS avdevice avformat swscale
-                                        avcodec avutil swresample)
-message(STATUS "FFMPEG_LIBRARIES: ${FFMPEG_LIBRARIES}")
-message(STATUS "FFMPEG_INCLUDE_DIRS: ${FFMPEG_INCLUDE_DIRS}")
+set(FFMPEG_ROOT ${AUTOCRANE_DEPS})
+find_package(FFMPEG 4.4 EXACT REQUIRED COMPONENTS avdevice avformat swscale
+    avcodec avutil swresample)
+message(STATUS " FFMPEG_LIBRARIES: ${FFMPEG_LIBRARIES} ")
+message(STATUS " FFMPEG_INCLUDE_DIRS: ${FFMPEG_INCLUDE_DIRS} ")
```
Comment (themightyoarfish) [inline CMakeLists.txt:39]:
```suggestion
message(STATUS "FFMPEG_LIBRARIES: ${FFMPEG_LIBRARIES}")
message(STATUS "FFMPEG_INCLUDE_DIRS: ${FFMPEG_INCLUDE_DIRS}")
```

File: CMakeLists.txt [hunk 20]
```diff
@@ -1,61 +1,207 @@
...
+
+# find_package(OpenMP REQUIRED)
+# set(CMAKE_CXX_FLAGS " ${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS} -DWITH_OPENMP ")
+# message(STATUS " OpenMP: ${OpenMP_CXX_LIBRARIES} ${OpenMP_INCLUDE_DIRS} ")
+# set(tclap_ROOT " /usr/local/Cellar/")
+# find_package(tclap REQUIRED PATHS " /usr/local/Cellar")
+# message(STATUS " tclap headers at ${tclap_INCLUDE_DIR} ")
+
+# tclap
+if(OSX)
```
Comment (themightyoarfish) [inline CMakeLists.txt:55]: so only compilable on osx now?

File: CMakeLists.txt [hunk 21]
```diff
@@ -1,61 +1,207 @@
...
+${CMAKE_CURRENT_LIST_DIR}/avutils.hpp
+)
+
+LIST(APPEND FILES
+  ${STREAM_SRCS}
+  ${STREAM_HEADERS})
+
+
+# define subgroups for XCode and other IDEs
+source_group( src FILES ${STREAM_HEADERS} ${STREAM_SRCS} )
```
Comment (themightyoarfish) [inline CMakeLists.txt:141]:
```suggestion
source_group(src FILES ${STREAM_HEADERS} ${STREAM_SRCS})
```

File: CMakeLists.txt (hunk 13 above)
Comment (themightyoarfish) [reply CMakeLists.txt:8]: guess we are now.

**Review by themightyoarfish — COMMENTED** [review]
it wouldve been easier to review if the changes (install stuff and addition of the encoding program and formatting) had been made in separate PRs). it seems to me this now assumes that its always used as an autocrane-core submodule, which is a new limitation.

File: CMakeLists.txt (hunk 14 above)
Comment (joelosw) [reply CMakeLists.txt:18]: Good Question, I guess for finding OpenCV on a windows machine? 
Not sure if we need windows-functionality at all though...

File: CMakeLists.txt (hunk 17 above)
Comment (joelosw) [reply CMakeLists.txt:25]:
```suggestion

```

File: CMakeLists.txt (hunk 20 above)
Comment (joelosw) [reply CMakeLists.txt:55]: On ubuntu tclap is found, on OSX we had some problems, which were fixed by Alex. 
So I guess it is okay, to only add the tclap-logic for OSX?

**Comment by joelosw** [comment]
> it wouldve been easier to review if the changes (install stuff and addition of the encoding program and formatting) had been made in separate PRs). it seems to me this now assumes that its always used as an autocrane-core submodule, which is a new limitation.

yes sorry about that.  It generically grew, I was not aware that we would change so much CMake logic

File: CMakeLists.txt (hunk 15 above)
Comment (joelosw) [reply CMakeLists.txt:22]: Was included from [your upstream](https://github.com/themightyoarfish/libffmpeg-zmq-streaming/commit/a4b5971340f79a6f8c8ca4b9db69fde4dd5180d4)

File: CMakeLists.txt (hunk 14 above)
Comment (themightyoarfish) [reply CMakeLists.txt:18]: then it can go away

File: CMakeLists.txt (hunk 15 above)
Comment (themightyoarfish) [reply CMakeLists.txt:22]: wtf i have no idea why thats there

File: CMakeLists.txt [hunk 22]
```diff
@@ -1,61 +1,207 @@
...
 find_package(OpenCV COMPONENTS core highgui imgcodecs imgproc REQUIRED)
+
 # Caution: due to either buggyness or horrendous design of gcc/ld/whatever, the order of
 # the libraries is important for linking: if lib A depends on stuff from lib B, lib B
 # *must* come after lib A. it does not help if lib B has already been passed previously
-find_package(FFMPEG REQUIRED COMPONENTS avdevice avformat swscale
-                                        avcodec avutil swresample)
-message(STATUS "FFMPEG_LIBRARIES: ${FFMPEG_LIBRARIES}")
-message(STATUS "FFMPEG_INCLUDE_DIRS: ${FFMPEG_INCLUDE_DIRS}")
+set(FFMPEG_ROOT ${AUTOCRANE_DEPS})
```
Comment (themightyoarfish) [inline CMakeLists.txt:34]: more uses of autocrane core deps

File: CMakeLists.txt (hunk 20 above)
Comment (themightyoarfish) [reply CMakeLists.txt:54]: what is the problem on osx? if this proj needs fixes to find tclap, it should also have a copy of the findtclap.cmake we made for the core repo.

Comment (joelosw) [reply CMakeLists.txt]: Right, done 👍

File: CMakeLists.txt (hunk 15 above)
Comment (joelosw) [reply CMakeLists.txt]: I removed it then....

**Review by themightyoarfish — APPROVED** [review]
It works. Should have a usage example though.

https://github.com/psiori/autocrane-core/pull/2418
